pkg/auth/providers/activedirectory/activedirectory_client.go

```diff
@@ -147,6 +165,43 @@ func (p *adProvider) RefetchGroupPrincipals(principalID string, secret string) (
 	return groupPrincipals, err
 }
 
+func getUsernameFromObjectGUID(lConn *ldapv3.Conn, config *v32.ActiveDirectoryConfig, username string) (string, error) {
+	uuid := strings.TrimPrefix(username, "objectGUID=")
+
+	parsedGUID, err := guid.Parse(uuid)
+	if err != nil {
+		return "", fmt.Errorf("parsing objectGUID to get the username: %w", err)
+	}
+
+	// search the user by its objectGUID
+	query := fmt.Sprintf(
+		"(&(%v=%v)(%s=%s))",
+		ObjectClass, config.UserObjectClass,
+		ObjectGUIDAttribute, guid.Escape(parsedGUID),
+	)
+	logrus.Debugf("LDAP Search query with objectGUID from UUID (%s): {%s}", parsedGUID.UUID(), query)
+
+	search := ldap.NewWholeSubtreeSearchRequest(
+		config.UserSearchBase,
+		query,
+		config.GetUserSearchAttributes(defaultUserAttributes...),
+	)
+
+	result, err := lConn.Search(search)
+	if err != nil {
+		return "", fmt.Errorf("LDAP search of user by objectGUID failed: %w", err)
+	}
+
+	if len(result.Entries) == 0 {
+		return "", errors.New("LDAP search of user by objectGUID returned no results")
+	} else if len(result.Entries) > 1 {
+		return "", fmt.Errorf("LDAP search of user by objectGUID returned no results")
```

Comment:
```suggestion
		return "", errors.New("LDAP search of user by objectGUID returned more than 1 result")
```

Could also be nice for both of these to include the user for debugging purposes,
```suggestion
		return "", fmt.Errorf("LDAP search of user %s by objectGUID returned more than 1 result: %v", username, result.Entries)
```

Reply:
_Could_ this leak personal information into logs?

Reply:
Not sure how sensitive could be, but since it's scoped to a login it should be easy enough to know about the problematic user, if needed.